### PR TITLE
Update boundary conditions to Bassi and Rebay (1997)

### PR DESCRIPTION
**Pull request by jkozdon**
- [x] Advection-Diffusion model
- [x] Heat eqn
- [x] Atmos model
- [x] Ocean model
- [x] poisson test

### CODE REVIEW

**Comment by jkozdon** [comment]
Had to create new PR, since PR #656 got closed automatically when branch jek/flux was merged (had the PR set to that branch to get a better diff). Could not reopen #656, hence this new PR.

**Comment by jkozdon** [comment]
bors try

**Comment by simonbyrne** [comment]
I mentioned to @fxgiraldo on the call this morning, our biggest challenge is that we mostly want to apply boundary conditions (particularly diffusive ones) directly by specifying the boundary flux.

[Something like this example](https://mathematica.stackexchange.com/a/127425/37818) might be a good test case?

**Comment by jkozdon** [comment]
bors try

**Comment by jkozdon** [comment]
bors try

**Comment by jkozdon** [comment]
bors try-

**Comment by jkozdon** [comment]
bors try

**Comment by jkozdon** [comment]
bors try

**Comment by simonbyrne** [comment]
This is looking really good, thanks!

File: src/DGmethods/NumericalFluxes.jl [hunk 1]
```diff
@@ -252,19 +238,33 @@ function numerical_boundary_flux_diffusive!(nf::CentralNumericalFluxDiffusive,
...
-                           Grad{S}(F⁻), state⁻, diff⁻, aux⁻,
+                           Grad{S}(F),
+                           state⁺, diff⁺, aux⁺,
+                           n⁻,
+                           state⁻, diff⁻, aux⁻,
                            bctype, t,
                            state1⁻, diff1⁻, aux1⁻)
 
-  Fᵀn .+= (F⁻ + F⁺)' * (n⁻/2)
+  Fᵀn .+= F' * n⁻
```
Comment (simonbyrne) [inline src/DGmethods/NumericalFluxes.jl]: Can we update the design docs to reflect this change?

Comment (simonbyrne) [reply src/DGmethods/NumericalFluxes.jl]: cc @fxgiraldo

File: src/DGmethods/NumericalFluxes.jl [hunk 2]
```diff
@@ -252,19 +238,33 @@ function numerical_boundary_flux_diffusive!(nf::CentralNumericalFluxDiffusive,
...
 
-  F⁻ = similar(Fᵀn, Size(3, nstate))
-  fill!(F⁻, -zero(FT))
-  flux_diffusive!(bl, Grad{S}(F⁻), state⁻, diff⁻, aux⁻, t)
-
-  F⁺ = similar(Fᵀn, Size(3, nstate))
-  fill!(F⁺, -zero(FT))
+  F = similar(Fᵀn, Size(3, nstate))
+  fill!(F, -zero(FT))
   boundary_flux_diffusive!(nf, bl,
```
[the remark this answers is not in the document]
Comment (jkozdon) [reply src/DGmethods/NumericalFluxes.jl]: Wouldn't this just be overloading `numerical_boundary_flux_diffusive!` completely?

[the remark this answers is not in the document]
Comment (jkozdon) [reply src/DGmethods/NumericalFluxes.jl]: i.e., this is what `numerical_boundary_flux_diffusive!` sets.

[the remark this answers is not in the document]
Comment (simonbyrne) [reply src/DGmethods/NumericalFluxes.jl]: It might be worth inserting an extra function for dispatch purposes.

**Comment by jkozdon** [comment]
bors try

**Comment by jkozdon** [comment]
bors try

**Comment by jkozdon** [comment]
bors try-

**Comment by jkozdon** [comment]
bors try

**Comment by jkozdon** [comment]
bors try

**Comment by jkozdon** [comment]
bors try

**Comment by jkozdon** [comment]
@blallen Can you grab the branch for this PR and make sure that all your ocean simulations look OK? I had to update the ocean model with the new boundary condition treatment.

**Comment by simonbyrne** [comment]
@jkozdon what's the status of this?

File: src/Atmos/Model/boundaryconditions.jl [hunk 3]
```diff
@@ -71,12 +76,17 @@ Set the momentum at the boundary to be zero.
...
-                               auxM::Vars, bctype, t, _...)
+function atmos_boundary_state!(nf::Union{Rusanov, CentralNumericalFluxGradient},
+                               bc::NoFluxBC, m::AtmosModel, stateP::Vars,
+                               auxP::Vars, nM, stateM::Vars, auxM::Vars, bctype,
+                               t, _...)
   FT = eltype(stateM)
   stateP.ρ = stateM.ρ
-  stateP.ρu -= 2 * dot(stateM.ρu, nM) * SVector(nM)
+  if typeof(nf) == Rusanov
+    stateP.ρu -= 2 * dot(stateM.ρu, nM) * SVector(nM)
```
Comment (simonbyrne) [inline src/Atmos/Model/boundaryconditions.jl]: why is this necessary?

Comment (jkozdon) [reply src/Atmos/Model/boundaryconditions.jl]: This is because we run this non-diffussive flux back through the numerical flux function, wheras for the Gradient terms it is used directly in the flux function.

@fxgiraldo can you confirm this is right?

**Comment by jkozdon** [comment]
> @jkozdon what's the status of this?

We have a weird Float32 issue difference between the slurm CI and our local system that I am trying to sort out. I have an idea, but haven't been able to test it. (We can comment out these weird results from the tests and open an issue if that's better)

Also would like @blallen (or someone else) to confirm the changes I made to the ocean morel are correct.

Otherwise this should be good to go.

File: src/Atmos/Model/boundaryconditions.jl (hunk 3 above)
Comment (simonbyrne) [reply src/Atmos/Model/boundaryconditions.jl]: Do we need this special handling for any other variables, or just momentum?

Comment (jkozdon) [reply src/Atmos/Model/boundaryconditions.jl]: Just momentum for this boundary condition type

Can you confirm @fxgiraldo

**Comment by jkozdon** [comment]
bors try

**Comment by jkozdon** [comment]
bors try

**Comment by jkozdon** [comment]
Now `Float32` is working for all tests. From looking at the [latest] slurm results and [previous] slurm results, suggests the the issue was Julia version 1.2 versus 1.3; Slurm CI upgraded to with climate-machine/slurmci#11

[latest]: https://gist.github.com/climabot/351b30191d17dc9dfe20d649caac0049
[previous]: https://gist.github.com/climabot/a9718f0db9d2d5475519b6d4631fd2f1
[PR]: https://github.com/climate-machine/slurmci/pull/11

**Comment by jkozdon** [comment]
bors try

**Comment by jkozdon** [comment]
Would be nice is someone (@kpamnany, @simonbyrne, @charleskawczynski, @smarras79?) could confirm that this PR does not break Dycoms; based on #664 it seems to be working now(?).

cc @blallen

**Comment by kpamnany** [comment]
DYCOMS appears to be running fine for this PR (from the SlurmCI try above: https://gist.github.com/climabot/d48549f5727f9f35ceb433ac98e981d8#file-out_8654789).

**Comment by jkozdon** [comment]
bors try

**Comment by jkozdon** [comment]
This is ready to go from my perspective.

**Comment by fxgiraldo** [comment]
> DYCOMS appears to be running fine for this PR (from the SlurmCI try above: https://gist.github.com/climabot/d48549f5727f9f35ceb433ac98e981d8#file-out_8654789).

Has someone confirmed that DYCOMS is running fine?

**Comment by jkozdon** [comment]
> > DYCOMS appears to be running fine for this PR (from the SlurmCI try above: https://gist.github.com/climabot/d48549f5727f9f35ceb433ac98e981d8#file-out_8654789).
> 
> Has someone confirmed that DYCOMS is running fine?

Currently the CI passed, but that just means that the code [didn't crash](https://github.com/climate-machine/CLIMA/blob/461657afa53ad608015a27f7c91dd8c15e605810/experiments/LES/dycoms.jl#L154-L185) since there is no `@test` in the test file.

**Comment by kpamnany** [comment]
Right. I'm hoping @akshaysridhar or @smarras79 can suggest a useful `@test` for the current 100sec simulation at some point.

**Comment by fxgiraldo** [comment]
> Right. I'm hoping @akshaysridhar or @smarras79 can suggest a useful `@test` for the current 100sec simulation at some point.

We need something to compare against when we perform the time-integrator study to make sure we know the stability condition parameters for the inviscid and viscous processes. Having a working DYCOMS up to the point that the stability measures (e.g., horizontal and vertical Courant numbers and Peclet numbers) equilibrate will be helpful.

**Comment by smarras79** [comment]
> > Right. I'm hoping @akshaysridhar or @smarras79 can suggest a useful `@test` for the current 100sec simulation at some point.
> 
> We need something to compare against when we perform the time-integrator study to make sure we know the stability condition parameters for the inviscid and viscous processes. Having a working DYCOMS up to the point that the stability measures (e.g., horizontal and vertical Courant numbers and Peclet numbers) equilibrate will be helpful.

Since we now have a way to I/O restart files, one thing that could be good to test only 100s of DYCOMS is to start the test from a restart file in the middle of the simulation so that all the initial transient is gone. Run it for 100s, and look at the norms difference for that slab of the run.
@akshaysridhar what do you think of this option?

**Comment by jkozdon** [comment]
This is likely a discussion to move to Issue #652

**Comment by blallen** [comment]
From a preliminary look, the ocean boundary conditions look good to me. I need to take a deeper look later today, but you can go ahead and merge this and I'll just make a small fix PR if things aren't right.

**Comment by jkozdon** [comment]
Thanks @blallen.

@simonbyrne should this be merged then?

**Comment by simonbyrne** [comment]
While I don't have any problem with the changes, I still don't think this addresses the core problem in that it is still somewhat unclear how they are supposed to be used, e.g. https://github.com/climate-machine/CLIMA/pull/660/files#r373245511.

That said, given the complexity of this PR, I don't want to hold it up, so if you feel that those points can be addressed in the design docs, i'm happy to merge it.

**Comment by blallen** [comment]
piggybacking off @simonbyrne :

my interpretation of this change is that we've switched from specific numerical fluxes (Rusanov, CentralNonDiff, CentralDiff, NumericalGradPenalty) to a generic NumericalFlux (either Diff or NonDiff) where the user defines both the numerical flux at the element interfaces and along the domain boundary (through two separate functions, but one calls the other). At the same time, additional methods were added so that this is backward compatible and you can still define boundary conditions in terms of ghost points instead of the expected flux.

is this generally correct?

**Comment by jkozdon** [comment]
The main goal was the bring the BC treatment in line with a [published method](https://doi.org/10.1006/jcph.1996.5572). The main changes are:

- for the gradient flux on boundaries, instead of calling the `numerical_flux_gradient!` with the minus side state and plus side (ghost) state, we just call `gradvariables!` and `diffusive!` on the plus side (ghost state); see [here](https://github.com/climate-machine/CLIMA/blob/d808788447a2354d1e5927b986d212780f078278/src/DGmethods/NumericalFluxes.jl#L49-L63)
- for the diffusive flux on the boundary we no longer call the `numerical_flux_diffussive!` on the minus and plus side states, but instead just call `flux_diffusive!` on the plus side (ghost) state; see [here](https://github.com/climate-machine/CLIMA/blob/d808788447a2354d1e5927b986d212780f078278/src/DGmethods/NumericalFluxes.jl#L241-L279)
- for the diffusive flux on the boundary it should be easier to specify the flux directly as opposed to having to define a plus side (ghost) state; see [here](https://github.com/climate-machine/CLIMA/blob/d808788447a2354d1e5927b986d212780f078278/src/DGmethods/NumericalFluxes.jl#L230-L263)

The treatment of the non-diffusive flux should be largely unchanged (i.e., we still define a plus side state and call the numerical flux).

If folks do not think this is useful, I'm fine abandoning the PR until we understand better what is needed.

**Comment by blallen** [comment]
Okay, that makes sense to me. For second order/parabolic terms, we only use the plus side when constructing the numerical flux on the boundary, while for first order/hyperbolic terms, we use both the plus and minus sides. 

I'll take a look through the paper to review everything and then see how the ocean boundary conditions fit into it. But it should be pretty straight forward to adapt to.

**Comment by simonbyrne** [comment]
Thanks @jkozdon for the summary, that does make it much clearer, and I do think this is a much improved interface for working with diffusive fluxes, which is certainly a current pain point.

My remaining concern is the handling of non-diffusive and gradient fluxes: e.g. it is still not clear to me how these conditions are derived, their relation to the Bassi & Rebay paper (e.g. the paper seems to one should use the same plus side for both?), or how we should prescribe alternatives (which I believe we will need for the LES case). But this sounds like it requires more input from @fxgiraldo, so I'm happy to merge this for now and try to address these other points later.
